[OP] WS don't retry on read error

The operator can panic since gorilla does not allow for repeated reads on websocket connections that returned an error on a read call, resulting in this:  
```
panic: repeated read on failed websocket connection

goroutine 49 [running]:
github.com/gorilla/websocket.(*Conn).NextReader(0xc00018fa20, 0xc000287af0, 0x470d9f, 0xecf258, 0xc000088190, 0x1313)
        /home/erdop/go/pkg/mod/github.com/gorilla/websocket@v1.4.1-0.20190629185528-ae1634f6a989/conn.go:965 +0x347
github.com/gorilla/websocket.(*Conn).ReadMessage(0xc00018fa20, 0xc000000002, 0xc000287bd8, 0x1, 0x1, 0xc0004897a0, 0x15e36d0)
        /home/erdop/go/pkg/mod/github.com/gorilla/websocket@v1.4.1-0.20190629185528-ae1634f6a989/conn.go:1026 +0x2f
github.com/perun-network/erdstall/operator.(*Peer).readMessages(0xc000485080)
        /home/erdop/erdstall/operator/rpc.go:156 +0x55
created by github.com/perun-network/erdstall/operator.(*RPCServer).connectionHandler
        /home/erdop/erdstall/operator/rpc.go:147 +0x427
```

Fixing this would mean to just remove the peer on the first read error, since gorilla does not seem to able to recover from it.  
Affects `operator/rpc.go Peer.readMessages`.

--- a/client/rpc.go
+++ b/client/rpc.go
@@ -67,11 +67,12 @@ func NewRPC(host string, port uint16) (*RPC, error) {
 	if err := conn.SetReadDeadline(time.Time{}); err != nil {
 		return nil, err
 	}
-	conn.SetCloseHandler(func(code int, err string) error {
-		rpc.Log().WithField("code", code).Errorf("WS connection was closed: %s", err)
-		return rpc.Close()
-	})
-	go rpc.handleConnections()
+	go func() {
+		err := rpc.handleConnections()
+		rpc.Log().WithError(err).Debug("RPC connection handler returned.")
+		err = rpc.Close()
+		rpc.Log().WithError(err).Debug("Stopped RPC client.")
+	}()
 
 	return rpc, nil
 }
@@ -142,14 +143,12 @@ func (r *RPC) Subscribe(ctx context.Context, user common.Address) (*Subscription
 	}
 }
 
-func (r *RPC) handleConnections() {
+func (r *RPC) handleConnections() error {
 	for !r.IsClosed() {
 		// gorilla has no async read method?!
 		_, data, err := r.conn.ReadMessage()
 		if err != nil {
-			r.Log().Error("reading ws message: ", err)
-			time.Sleep(1 * time.Second)
-			continue
+			return fmt.Errorf("reading ws message: %w", err)
 		}
 		r.Log().Trace("client received: ", string(data))
 		var msg wire.Result
@@ -164,13 +163,13 @@ func (r *RPC) handleConnections() {
 		case msg.Topic != "":
 			if r.subscription == nil {
 				r.Log().Error("Received proof without subscription")
-				return
 			}
 			r.subscription.handleTopic(msg.Topic, data)
 		default:
 			r.Log().Error("Received result without ID or Topic")
 		}
 	}
+	return nil
 }
 
 func (r *RPC) registerCallback(id wire.ID, cb callback) {

--- a/client/rpc_test.go
+++ b/client/rpc_test.go
@@ -134,4 +134,7 @@ func TestRPC_ClientOp(t *testing.T) {
 		require.NoError(t, err)
 		assert.Equal(t, *dp, proof)
 	})
+
+	assert.NoError(t, rpcClient.Close())
+	assert.NoError(t, rpcServer.Close())
 }

--- a/operator/rpc.go
+++ b/operator/rpc.go
@@ -139,25 +139,26 @@ func (r *RPCServer) connectionHandler(out http.ResponseWriter, in *http.Request)
 	peer := &Peer{conn: conn, op: r.op}
 	if err := peer.sendJSON(r.server.clientConfig); err != nil {
 		r.Log().WithError(err).Error("Pushing ClientConfig")
+		return
 	}
-
-	conn.SetCloseHandler(func(int, string) error {
-		return peer.Close()
-	})
-	go peer.readMessages()
+	// Start client handler routine.
+	go func() {
+		err := peer.readMessages()
+		r.Log().WithError(err).Debug("Peer connection handler returned.")
+		err = peer.Close()
+		r.Log().WithError(err).Debug("Stopped Peer.")
+	}()
 }
 
 func (p *Peer) Log() *log.Entry {
 	return log.WithField("role", "peer")
 }
 
-func (p *Peer) readMessages() {
+func (p *Peer) readMessages() error {
 	for !p.IsClosed() {
 		_, msg, err := p.conn.ReadMessage()
 		if err != nil {
-			p.Log().WithError(err).Error("Could not read msg")
-			time.Sleep(1 * time.Second)
-			continue
+			return fmt.Errorf("reading ws message: %w", err)
 		}
 
 		var call wire.Call
@@ -171,6 +172,7 @@ func (p *Peer) readMessages() {
 			p.Log().WithField("id", call.ID).WithError(sendErr).Error("Could not send result")
 		}
 	}
+	return nil
 }
 
 func (p *Peer) Close() error {

